src/dx/subscribe/event_engine/event.rs

```diff
@@ -0,0 +1,151 @@
+    /// [`PubNub`] network (network issues or revoked / expired permissions).
+    ///
+    /// [`PubNub`]: https://www.pubnub.com/
+    ReceiveFailure { reason: PubNubError },
+
+    /// Receive updates reconnect completed successfully.
+    ///
+    /// Emitted when another receive updates attempt was successful and
+    /// [`PubNub`] network returned list of real-time updates along
+    /// timetoken (cursor) which will be used for subscription loop.
+    ///
+    /// [`PubNub`]: https://www.pubnub.com/
+    ReceiveReconnectSuccess {
+        cursor: SubscribeCursor,
+        messages: Vec<String>,
+    },
+
+    /// Receive updates reconnect completed with error.
+    ///
+    /// Emitted when another receive updates effect attempt was unable to
+    /// receive response from [`PubNub`] network (network issues or
+    /// revoked permissions).
+    ///
+    /// [`PubNub`]: https://www.pubnub.com/
+    ReceiveReconnectFailure { reason: PubNubError },
+
+    /// All receive updates attempts was unsuccessful.
+    ///
+    /// Emitted when receive updates reconnect attempts reached maximum allowed
+    /// count (according to retry / reconnection policy) and all following
+    /// attempts should be stopped.
+    ReceiveReconnectGiveUp { reason: PubNubError },
+
+    /// Disconnect from [`PubNub`] network.
+    ///
+    /// Emitted when explicitly requested to stop receiving real-time updates.
+    ///
+    /// [`PubNub`]: https://www.pubnub.com/
+    Disconnect,
+
+    /// Reconnect to [`PubNub`] network.
+    ///
+    /// Emitted when explicitly requested to restore real-time updates receive.
+    ///
+    /// [`PubNub`]: https://www.pubnub.com/
+    Reconnect,
+}
+
+impl Event for SubscribeEvent {
+    fn id(&self) -> &str {
+        match self {
+            SubscribeEvent::SubscriptionChanged { .. } => "SUBSCRIPTION_CHANGED",
+            SubscribeEvent::SubscriptionRestored { .. } => "SUBSCRIPTION_RESTORED",
+            SubscribeEvent::HandshakeSuccess { .. } => "HANDSHAKE_SUCCESS",
+            SubscribeEvent::HandshakeFailure { .. } => "HANDSHAKE_FAILURE",
+            SubscribeEvent::HandshakeReconnectSuccess { .. } => "HANDSHAKE_RECONNECT_SUCCESS",
+            SubscribeEvent::HandshakeReconnectFailure { .. } => "HANDSHAKE_RECONNECT_FAILURE",
+            SubscribeEvent::HandshakeReconnectGiveUp { .. } => "HANDSHAKE_RECONNECT_GIVEUP",
+            SubscribeEvent::ReceiveSuccess { .. } => "RECEIVE_SUCCESS",
+            SubscribeEvent::ReceiveFailure { .. } => "RECEIVE_FAILURE",
+            SubscribeEvent::ReceiveReconnectSuccess { .. } => "RECEIVE_RECONNECT_SUCCESS",
+            SubscribeEvent::ReceiveReconnectFailure { .. } => "RECEIVE_RECONNECT_FAILURE",
+            SubscribeEvent::ReceiveReconnectGiveUp { .. } => "RECEIVE_RECONNECT_GIVEUP",
+            SubscribeEvent::Disconnect => "DISCONNECT",
+            SubscribeEvent::Reconnect => "RECONNECT",
+        }
```

Comment:
I think `Display` trait is made for that usage.

Reply:
Had some discussion about this and, at least for now, decided to keep as is (there is no need to make formatted messages with `Display`, just plain strings).